### PR TITLE
Add ufs driver. Remove NEMS submodule. Update CMEPS. Update CDEPS. Feature/ort addition. 

### LINKED ISSUE #59: Write ATM history files as cubed-sphere tiles instead of Mesh
When mediator history files are enabled for CMEPS, the ATM component is written on the ATM mesh. This limits the usefulness of the history files because an additional processing step is required to place the fields on the ATM tiled grid for visualization.

Allowing the ATM history files to be optionally written on the cubed sphere tiles would allow developers to make quick checks of coupling fields during development. It could also be useful when implementing a set of graphical outputs that could be generated along with the RTs, as suggested in 
- https://github.com/ufs-community/ufs-weather-model/issues/914

.

### LINKED ISSUE #930: Remove NEMS submodule
ufs-weather-model currently uses two Fortran source files from the NEMS repository. We can remove dependency on that repository and delete NEMS submodule if we move those two files to this repository.
See this discussion: https://github.com/ufs-community/ufs-weather-model/discussions/872

### PR DESCRIPTION
**Pull request by DusanJovic-NOAA**
# PR Checklist

- [x] Ths PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [x] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [ ] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR 
are specified below.

- [ ] Results for one or more of the regression tests change and the reasons for the changes are understood and explained below.

- [ ] New or updated input data is required by this PR. If checked, please work with the code managers to update input data sets on all platforms.

## Instructions: All subsequent sections of text should be filled in as appropriate.

The information provided below allows the code managers to understand the changes relevant to this PR, whether those changes are in the ufs-weather-model repository or in a subcomponent repository. Ufs-weather-model code managers will use the information provided to add any applicable labels, assign reviewers and place it in the Commit Queue. Once the PR is in the Commit Queue, it is the PR owner's responsiblity to keep the PR up-to-date with the develop branch of ufs-weather-model. 

## Description

This PR adds UFS driver, a copy of two fortran files from the NEMS repository, and removes NEMS submodule.
The following three PRs have been combined into this one: #932, #939 and #928 

### Issue(s) addressed

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues must always be created before starting work on a PR branch!) 
- fixes #930
- fixes https://github.com/NOAA-EMC/CMEPS/issues/59

## Testing

How were these changes tested? Yes.
What compilers / HPCs was it tested with? Intel on Hera.
Are the changes covered by regression tests? Yes.
Have regression tests and unit tests (utests) been run? On which platforms and with which compilers? (Note that unit tests can only be run on tier-1 platforms)

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel 
- [x] cheyenne.gnu
- [x] gaea.intel 
- [x] jet.intel
- [x] wcoss_cray
- [x] wcoss_dell_p3
- [ ] opnReqTest for newly added/changed feature
- [x] CI https://github.com/ufs-community/ufs-weather-model/actions/runs/1584662830

## Dependencies

If testing this branch requires non-default branches in other repositories, list them. Those branches should have matching names (ideally).

Do PRs in upstream repositories need to be merged first? 
- waiting on https://github.com/NOAA-EMC/CMEPS/pull/60

### CODE REVIEW

**Comment by junwang-noaa** [comment]
Can we put the UFS.F90 and the EARTH_GRID_COMP.F90 in driver/ directory?

**Comment by DusanJovic-NOAA** [comment]
> Can we put the UFS.F90 and the EARTH_GRID_COMP.F90 in driver/ directory?

Done. In this commit https://github.com/ufs-community/ufs-weather-model/pull/931/commits/876109732f0bbc392493e2a39318a6c400ad3ee5

File: tests/module-setup.sh [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    fi
+    module purge
+
+elif [[ $MACHINE_ID = wcoss_cray ]] ; then
+    # We are on NOAA Luna or Surge
+    if ( ! eval module help > /dev/null 2>&1 ) ; then
+        source /opt/modules/default/init/bash
+    fi
+    module purge
+    module purge
```
Comment (aerorahul) [inline tests/module-setup.sh]:
```suggestion
```

File: tests/module-setup.sh [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        # before loading the module command.  Without those variables,
+        # the module command fails.  Hence we actually have to source
+        # /etc/profile here.
+        source /etc/profile
+        __ms_source_etc_profile=yes
+    else
+        __ms_source_etc_profile=no
+    fi
+    module purge
+# clean up after purge
```
Comment (aerorahul) [inline tests/module-setup.sh]:
```suggestion
    # clean up after purge
```

File: tests/module-setup.sh [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    # We are on NOAA Mars or Venus
+    if ( ! eval module help > /dev/null 2>&1 ) ; then
+        source /usrx/local/prod/lmod/lmod/init/bash
+    fi
+    module purge
+
+elif [[ $MACHINE_ID = cheyenne* ]] ; then
+    # We are on NCAR Cheyenne
+    if ( ! eval module help > /dev/null 2>&1 ) ; then
+        . /glade/u/apps/ch/modulefiles/default/localinit/localinit.sh
```
Comment (aerorahul) [inline tests/module-setup.sh]:
```suggestion
        source /glade/u/apps/ch/modulefiles/default/localinit/localinit.sh
```

**Review by aerorahul — COMMENTED** [review]
Minor touch ups.

File: driver/UFS.F90 [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+                     ,localpet=MYPE                                     &  !<-- The local MPI task ID
+                     ,rc      =RC)
+      ESMF_ERR_ABORT(RC)
+! ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
+!
+!-----------------------------------------------------------------------
+!***  Print subversion version and other status information.
+!-----------------------------------------------------------------------
+!
+      if (mype==0) call w3tagb('nems     ',0000,0000,0000,'np23   ')
```
Comment (aerorahul) [inline driver/UFS.F90]: This is no longer `nems`.  Similar comment at the end with `w3tage`

File: driver/UFS.F90 [hunk 5]
```diff
@@ -0,0 +1,515 @@
+#define ESMF_ERR_ABORT(rc) if (ESMF_LogFoundError(rc, msg="Aborting NEMS", line=__LINE__, file=__FILE__)) call ESMF_Finalize(endflag=ESMF_END_ABORT)
+!-----------------------------------------------------------------------
+!
+      PROGRAM UFS
```
Comment (aerorahul) [inline driver/UFS.F90:4]: Is there a reason why we keep using  fixed format in our codes.  This is not F77 anymore.

File: driver/UFS.F90 [hunk 6]
```diff
@@ -0,0 +1,515 @@
...
+
+!-----------------------------------------------------------------------
+!***  Shut down the ESMF system.
+!-----------------------------------------------------------------------
+!
+      CALL ESMF_Finalize()
+!
+!-----------------------------------------------------------------------
+!
+      if(mype==0)call w3tage('nems     ')
```
Comment (aerorahul) [inline driver/UFS.F90]: Not nems any more.  Suggest replacing w/ `ufs`.

File: driver/UFS.F90 (hunk 5 above)
Comment (aerorahul) [reply driver/UFS.F90:4]: Same comment on the other file `driver/EARTH_GRID_COMP.F90` added.

**Review by aerorahul — CHANGES_REQUESTED** [review]
some comments about the formatting of the new code added and W3 tags.

File: driver/UFS.F90 (hunk 5 above)
Comment (DusanJovic-NOAA) [reply driver/UFS.F90:4]: It is not F77, this is F90. It's also in free format.

**Comment by junwang-noaa** [comment]
Did we test if the build.sh is working which is used outside RT?

**Comment by junwang-noaa** [comment]
Also we may need to add doxygen in the two Fortran files later.

**Comment by DusanJovic-NOAA** [comment]
> Did we test if the build.sh is working which is used outside RT?

It works:
```
 module use modulefiles
 module load ufs_hera.intel
 CMAKE_FLAGS="-DAPP=ATM -DCCPP_SUITES=FV3_GFS_v16" ./build.sh 
```

**Comment by MinsukJi-NOAA** [comment]
@DusanJovic-NOAA please make this one file change for CI repo check. This is required since NEMS submodule no longer exists.
```
diff --git a/tests/ci/repo_check.sh b/tests/ci/repo_check.sh
index e2cf8ec..b8d0487 100755
--- a/tests/ci/repo_check.sh
+++ b/tests/ci/repo_check.sh
@@ -13,8 +13,8 @@ result() {
 }
 
 # Declare variables
-declare -A base fv3 mom6 cice ww3 stoch gocart nems cmeps cdeps hycom cmake
-submodules="fv3 mom6 cice ww3 stoch gocart nems cmeps cdeps hycom cmake"
+declare -A base fv3 mom6 cice ww3 stoch gocart cmeps cdeps hycom cmake
+submodules="fv3 mom6 cice ww3 stoch gocart cmeps cdeps hycom cmake"
 comment=''
 ownerID=$1
 
@@ -47,10 +47,6 @@ gocart[repo]='https://github.com/GEOS-ESM/GOCART'
 gocart[branch]='develop'
 gocart[dir]='GOCART'
 
-nems[repo]='https://github.com/NOAA-EMC/NEMS'
-nems[branch]='develop'
-nems[dir]='NEMS'
-
 cmeps[repo]='https://github.com/NOAA-EMC/CMEPS'
 cmeps[branch]='emc/develop'
 cmeps[dir]='CMEPS-interface/CMEPS'
```

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: gaea
Compiler: intel
Job: RT
Repo location: /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/787280954/20211215213009/ufs-weather-model
Please manually delete: /lustre/f2/scratch/emc.nemspara/FV3_RT/rt_25118
Test compile 014 failed failed 
Please make changes and add the following label back:
gaea-intel-RT

**Comment by climbfuji** [comment]
I think the PR description/title need updates for all the other PRs that have been pulled into this one? At least links to them and notes that they have been included?

**Comment by DusanJovic-NOAA** [comment]
> _**Automated RT Failure Notification**_ Machine: gaea Compiler: intel Job: RT Repo location: /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/787280954/20211215213009/ufs-weather-model Please manually delete: /lustre/f2/scratch/emc.nemspara/FV3_RT/rt_25118 Test compile 014 failed failed Please make changes and add the following label back: gaea-intel-RT

Looks like job compile_014 finished successfully:

```
$ tail /lustre/f2/scratch/emc.nemspara/FV3_RT/rt_25118/compile_014/err
+ cp /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/787280954/20211215213009/ufs-weather-model/modulefiles/ufs_gaea.intel /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/787280954/20211215213009/ufs-weather-model/tests/modules.fv3_014
+ '[' YES = YES ']'
+ rm -rf /lustre/f2/scratch/emc.nemspara/FV3_RT/rt_25118/compile_014/build_fv3_014
+ elapsed=502
+ echo 'Elapsed time 502 seconds. Compiling -DAPP=ATMW -DCCPP_SUITES=FV3_GFS_v16 -D32BIT=ON -DMPI=ON -DCMAKE_BUILD_TYPE=Release finished'
+ echo 'Compile 014 elapsed time 502 seconds. -DAPP=ATMW -DCCPP_SUITES=FV3_GFS_v16 -D32BIT=ON -DMPI=ON -DCMAKE_BUILD_TYPE=Release'
++ date
+ echo 'Compile ended:    ' Wed Dec 15 22:17:39 UTC 2021
++ date +%s
+ echo -n ' 1639606659,'

```

**Comment by climbfuji** [comment]
I have seen this happen elsewhere in the last weeks. Is it looking for the executable? If so, then delays in the filesystem synchronization could be the problem.

**Comment by DusanJovic-NOAA** [comment]
> I have seen this happen elsewhere in the last weeks. Is it looking for the executable? If so, then delays in the filesystem synchronization could be the problem.
> […]

I see the executable copied from the build directory over to test directory:
```
$ ls -l /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/787280954/20211215213009/ufs-weather-model/tests/fv3_014.exe
-rwxr-xr-x 1 emc.nemspara ncep 319696744 Dec 15 17:17 /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/787280954/20211215213009/ufs-weather-model/tests/fv3_014.exe
```

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: jet
Compiler: intel
Job: RT
Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/787280954/20211215213013/ufs-weather-model
Please manually delete: /lfs4/HFIP/h-nems/emc.nemspara/RT_RUNDIRS/emc.nemspara/FV3_RT/rt_127275
Test control_csawmg 047 failed failed 
Test control_csawmg 047 failed in run_test failed 
Please make changes and add the following label back:
jet-intel-RT

**Review by junwang-noaa — APPROVED** [review]

**Review by aerorahul — APPROVED** [review]

**Review by binli2337 — APPROVED** [review]

File: tests/module-setup.sh [hunk 7]
```diff
@@ -0,0 +1,117 @@
+#!/bin/bash
```
Comment (climbfuji) [inline tests/module-setup.sh:1]: Thank you SO much for getting rid of this nasty "testing for certain directories to determine where I am" logic!

**Review by climbfuji — APPROVED** [review]

**Review by MinsukJi-NOAA — APPROVED** [review]

**Comment by binli2337** [comment]
The related CDEPS PR has been merged.

**Review by DeniseWorthen — APPROVED** [review]
CMEPS changes are good, hash is correct

**Pull request merged**